File: backend/app/services/ror.py
"""Institution coordinates and ids from the public ROR API (ror.org).

Backs the ``seed-coordinates`` CLI command. Parsing is kept separate from
the network fetches so it can be unit-tested against fixed JSON payloads
without network access. Coordinates come from the record's first location's
``geonames_details`` — the same field the institution edit form's
browser-side "Fetch from ROR" button reads.
"""

import re
from dataclasses import dataclass

import httpx

ROR_API = "https://api.ror.org/v2/organizations"
USER_AGENT = "USMCCDB (+https://github.com/trholmes/usmccdb)"


@dataclass
class RorMatch:
    ror_id: str  # bare form, e.g. "05gvnxz63"
    name: str
    latitude: float | None
    longitude: float | None
    short_name: str | None = None  # candidate short_name (see parse_short_name)
    address: str | None = None  # draft author-list (latex) address


def parse_coordinates(record: dict) -> tuple[float, float] | None:
    """First location's geonames coordinates of a v2 record, if present."""
    locations = record.get("locations") or [{}]
    geo = locations[0].get("geonames_details") or {}
    lat, lng = geo.get("lat"), geo.get("lng")
    if lat is None or lng is None:
        return None
    return float(lat), float(lng)


def parse_acronym(record: dict) -> str | None:
    """The record's acronym ("UTK", "FNAL", …)."""
    for name in record.get("names") or []:
        if "acronym" in (name.get("types") or []):
            return name.get("value") or None
    return None


def _university_short_name(name: str) -> str | None:
    """The distinctive part of a university name, the way collaboration
    lists abbreviate them: "Cornell University" → "Cornell", "University of
    Chicago" → "Chicago", "University of California, Berkeley" → "UC
    Berkeley". None when the name doesn't fit a pattern we trust."""
    name = re.sub(r"^The\s+", "", name.strip())
    m = re.match(r"^University of California[,–-]\s*(.+)$", name, re.IGNORECASE)
    if m:
        return f"UC {m.group(1)}"
    m = re.match(r"^University of (.+)$", name, re.IGNORECASE)
    if m and "," not in m.group(1):
        return m.group(1)
    m = re.match(r"^(.+?) University$", name, re.IGNORECASE)
    if m and " of " not in m.group(1).lower():
        return m.group(1)
    return None


def parse_short_name(record: dict) -> str | None:
    """Candidate short name for a record. Universities (ROR type
    "education") read better as the distinctive part of their name
    ("Cornell", not "CU"); labs and everything else keep their acronym
    (FNAL, BNL, …). Falls back to the acronym when the university name
    doesn't fit a known pattern (e.g. "Massachusetts Institute of
    Technology" → "MIT")."""
    if "education" in (record.get("types") or []):
        derived = _university_short_name(_display_name(record) or "")
        if derived:
            return derived
    return parse_acronym(record)


def parse_address(record: dict) -> str | None:
    """A draft author-list address: "<name>, <city>, <ST>, USA" for US
    records (ROR has no street/zip, so this is a starting point for the
    office to refine), "<name>, <city>, <country>" elsewhere."""
    display = _display_name(record)
    locations = record.get("locations") or [{}]
    geo = locations[0].get("geonames_details") or {}
    city = geo.get("name")
    if not display or not city:
        return None
    parts = [display, city]
    if geo.get("country_code") == "US":
        subdivision = geo.get("country_subdivision_code")
        if subdivision:
            parts.append(subdivision)
        parts.append("USA")
    elif geo.get("country_name"):
        parts.append(geo["country_name"])
    return ", ".join(parts)


def _bare_id(record: dict) -> str:
    # v2 record ids are full URLs like https://ror.org/05gvnxz63
    return str(record.get("id", "")).rstrip("/").rsplit("/", 1)[-1]


def _display_name(record: dict) -> str | None:
    for name in record.get("names") or []:
        if "ror_display" in (name.get("types") or []):
            return name.get("value") or None
    return None


def parse_affiliation_match(payload: dict) -> RorMatch | None:
    """The single confident ("chosen") match of an affiliation query, if any.

    ROR's affiliation matcher flags at most one item as chosen; anything
    less certain needs a human decision, so it is deliberately ignored here.
    """
    for item in payload.get("items") or []:
        if item.get("chosen"):
            org = item.get("organization") or {}
            coords = parse_coordinates(org)
            return RorMatch(
                ror_id=_bare_id(org),
                name=_display_name(org) or _bare_id(org),
                latitude=coords[0] if coords else None,
                longitude=coords[1] if coords else None,
                short_name=parse_short_name(org),
                address=parse_address(org),
            )
    return None


def fetch_record(client: httpx.Client, ror_id: str) -> dict:
    resp = client.get(f"{ROR_API}/{ror_id}")
    resp.raise_for_status()
    return resp.json()


def fetch_affiliation_match(client: httpx.Client, affiliation: str) -> RorMatch | None:
    resp = client.get(ROR_API, params={"affiliation": affiliation})
    resp.raise_for_status()
    return parse_affiliation_match(resp.json())

File: frontend/src/api/ror.ts
// Browser-side ROR (ror.org) lookups and record parsing, shared by the
// institution edit form and the bulk "Fill from ROR" flow. The lookups run
// in the admin's browser (not the backend), so an air-gapped server still
// works. Mirrors backend/app/services/ror.py.

const ROR_API = 'https://api.ror.org/v2/organizations'
const ID_RE = /(0[a-z0-9]{8})$/

/** The fields we can pull out of a ROR v2 organization record. */
export interface RorParsed {
  rorId: string | null
  name: string
  shortName: string | null // candidate short name (see parseRorRecord)
  address: string | null // draft author-list address
  latitude: number | null
  longitude: number | null
  location: string | null // "Knoxville, United States" — for display only
  isUS: boolean | null
  score: number | null // affiliation-matcher confidence, when it came from one
}

// The distinctive part of a university name, the way collaboration lists
// abbreviate them: "Cornell University" → "Cornell", "University of
// Chicago" → "Chicago", "University of California, Berkeley" → "UC
// Berkeley". null when the name doesn't fit a pattern we trust.
function universityShortName(name: string): string | null {
  name = name.trim().replace(/^The\s+/, '')
  let m = name.match(/^University of California[,–-]\s*(.+)$/i)
  if (m) return `UC ${m[1]}`
  m = name.match(/^University of (.+)$/i)
  if (m && !m[1].includes(',')) return m[1]
  m = name.match(/^(.+?) University$/i)
  if (m && !m[1].toLowerCase().includes(' of ')) return m[1]
  return null
}

export function parseRorRecord(rec: any, score: number | null = null): RorParsed {
  const names: any[] = rec.names ?? []
  const display: string | null =
    names.find((n) => n.types?.includes('ror_display'))?.value ?? names[0]?.value ?? null
  const acronym: string | null =
    names.find((n) => n.types?.includes('acronym'))?.value ?? null
  // Universities read better as the distinctive part of their name
  // ("Cornell", not "CU"); labs and everything else keep their acronym
  // (FNAL, BNL, …). Mirrors parse_short_name in backend/app/services/ror.py.
  const shortName =
    (rec.types?.includes('education') && display ? universityShortName(display) : null) ??
    acronym
  const geo = rec.locations?.[0]?.geonames_details ?? {}
  const city: string | null = geo.name ?? null
  // Same draft-address shape the backend builds: "Name, City, ST, USA" for
  // US records (ROR has no street/zip), "Name, City, Country" elsewhere.
  let address: string | null = null
  if (display && city) {
    const parts = [display, city]
    if (geo.country_code === 'US') {
      if (geo.country_subdivision_code) parts.push(geo.country_subdivision_code)
      parts.push('USA')
    } else if (geo.country_name) {
      parts.push(geo.country_name)
    }
    address = parts.join(', ')
  }
  return {
    rorId: String(rec.id ?? '').match(ID_RE)?.[1] ?? null,
    name: display ?? '(unnamed)',
    shortName,
    address,
    latitude: geo.lat ?? null,
    longitude: geo.lng ?? null,
    location: city ? `${city}${geo.country_name ? `, ${geo.country_name}` : ''}` : null,
    isUS: geo.country_code ? geo.country_code === 'US' : null,
    score,
  }
}

export async function fetchRorRecord(rorId: string): Promise<RorParsed> {
  const resp = await fetch(`${ROR_API}/${rorId}`)
  if (!resp.ok) throw new Error(`ROR lookup failed (${resp.status})`)
  return parseRorRecord(await resp.json())
}

export interface RorAffiliationResult {
  /** ROR's single confident match, if it flagged one. */
  chosen: RorParsed | null
  /** Best candidates (chosen or not) for a human to pick from. */
  candidates: RorParsed[]
}

/** Query ROR's affiliation matcher with a free-text institution string. */
export async function matchRorAffiliation(query: string): Promise<RorAffiliationResult> {
  const resp = await fetch(`${ROR_API}?affiliation=${encodeURIComponent(query)}`)
  if (!resp.ok) throw new Error(`ROR lookup failed (${resp.status})`)
  const items: any[] = (await resp.json()).items ?? []
  const chosen = items.find((it) => it.chosen)
  return {
    chosen: chosen ? parseRorRecord(chosen.organization, chosen.score ?? null) : null,
    candidates: items
      .slice(0, 5)
      .map((it) => parseRorRecord(it.organization, it.score ?? null)),
  }
}
